README.rst
===========
pmuse_utils
===========


This module is made of short scripts which would hopefully help dealing with PHANGS MUSE data

Description
===========

The pmuse_utils package is meant to provide a few functionalities to help addressing the PHANGS MUSE data and
build figures.


Note
====

This project has been set up using PyScaffold 3.2.3. For details and usage
information on PyScaffold see https://pyscaffold.org/.


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- Adding old python scripts
- Fixing importing and config


tests/conftest.py
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for pmuse_utils.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""

# import pytest


tests/test_skeleton.py
# -*- coding: utf-8 -*-

import pytest
from pmuse_utils.skeleton import fib

__author__ = "Eric Emsellem"
__copyright__ = "Eric Emsellem"
__license__ = "mit"


def test_fib():
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


src/pmuse_utils/skeleton.py
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
[options.entry_points] section in setup.cfg:

    console_scripts =
         fibonacci = pmuse_utils.skeleton:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""

import argparse
import sys
import logging

from pmuse_utils import __version__

__author__ = "Eric Emsellem"
__copyright__ = "Eric Emsellem"
__license__ = "mit"

_logger = logging.getLogger(__name__)


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n-1):
        a, b = b, a+b
    return a


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="pmuse_utils {ver}".format(ver=__version__))
    parser.add_argument(
        dest="n",
        help="n-th Fibonacci number",
        type=int,
        metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO)
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


src/pmuse_utils/phangs_muse_config.py
# -*- coding: utf-8 -*-

# ============== USER DEFINED PARAM =========================
# Version you wish to use for the input Data/Table
BB = "1.0"
DR = "2.1"
version_phangs_table = "v1p6"

# Where the init py script is on your computer
base_folder = "/home/science/PHANGS/"
muse_subfolder =  "MUSE/"
bb_subfolder = "BB_images/"
# Where is the "Info" folder
pyscript_folder = f"{base_folder}{muse_subfolder}pymuse_scripts/"
dr_folder = f"{base_folder}{muse_subfolder}data/DR{DR}/"

# Function setting up the folder dictionary - you can change
# the values in the dictionary according to your structure 
# but pls do not change the function itself
def define_phangs_keywords(base_folder=base_folder,
                           DR=DR, BB=BB, 
                           version_phangs_table=version_phangs_table):
    """Define the keywords needed for phangs
    """

    dict_phangs_keywords = {
            # From base folder, here is the set of folders used in this script
            "BB": BB,
            "DR": DR,
            "version_phangs_table": version_phangs_table,
            "phangs_data_folder" : f"{base_folder}PhangsTables/",
            "bb_folder" : f"{base_folder}{bb_subfolder}/DR{BB}/",
            "wfi_folder" : f"{base_folder}{bb_subfolder}/DR{BB}/WFI_images/",
            "dp_folder" : f"{base_folder}{bb_subfolder}/DR{BB}/DuPont_images/",
            "muse_folder": f"{base_folder}{muse_subfolder}",
            "dr_folder" : dr_folder,
            "mask_folder" : f"{base_folder}{muse_subfolder}data/DR{DR}/masks/",
            "muse_info_folder" : f"{pyscript_folder}Info/DR{DR}/",
            "native_folders": [f"{dr_folder}native_IMAGES/", f"{dr_folder}native_MAPS/"],
            "copt_folders": [f"{dr_folder}copt_IMAGES/", f"{dr_folder}copt_MAPS/"]
            }
    return dict_phangs_keywords

# ============= MAPS  =======================================
# For colours
default_rgb_bands = {'R': "SDSS_i", 'G': "SDSS_r", 'B': "SDSS_g"}
default_white = {'white': "white"}

default_colormap = "viridis"
def_perc = [2, 99]

# Rule for maps (cmap and stretch)
# Colors - please install it, cmocean is very nice!
# Just install it via:
#     pip install cmocean
import cmocean

# Using labels - first is stretch, second is color map, last is centred cuts or not
maps_rules_dict = {"V_STARS": [cmocean.cm.balance, 'linear', False, def_perc, True],
                   "SIGMA_STARS": [cmocean.cm.thermal, 'linear', False, [5, 97], False],
                   "AGE_LW": ['Greys_r', 'linear', False, [1, 99], False],
                   "AGE_MW": ['Greys_r', 'linear', False, def_perc, False],
                   "Z_LW": ['Greys_r', 'linear', False, def_perc, False],
                   "Z_MW": ['Greys_r', 'linear', False, def_perc, False]}

# Also keywords work - beware, this is not robust and it needs to be exclusive
maps_keyw_dict = {"_VEL": [cmocean.cm.balance, 'linear', False, def_perc, True],
                  "_SIGMA": [cmocean.cm.thermal, 'linear', False, [5, 99], False],
                  "_FLUX": [default_colormap, 'sqrt', False, [5, 99], False],
                  "WFI": [default_colormap, 'asinh', False, [2, 99.5], False]
                  }

dict_config_files = {'2.1': ["phangsdictionaries_v19", "psf_dictionary_v6"]}
# ============== END USER DEFINED PARAM =====================


src/pmuse_utils/dict/psf_dictionary_v6.py
# v6.0 - changed for DR2.1
# fwhm, alpha, lambda0, b

dict_psf_150pc = {'IC5332': 3.43393120433962,
                  'NGC0628': 3.1442805342769633,
                  'NGC1087': 1.952032819711232,
                  'NGC1300': 1.6292638920470441,
                  'NGC1365': 1.5809770781574026,
                  'NGC1385': 1.7967318333787632,
                  'NGC1433': 1.660747307075634,
                  'NGC1512': 1.6431078631133105,
                  'NGC1566': 1.7489949125494855,
                  'NGC1672': 1.594831007497124,
                  'NGC2835': 2.531892001716559,
                  'NGC3351': 3.1063976725466658,
                  'NGC3627': 2.733190969184465,
                  'NGC4254': 2.361810689779021,
                  'NGC4303': 1.821054817767559,
                  'NGC4321': 2.034169681753257,
                  'NGC4535': 1.9619353226989054,
                  'NGC5068': 5.9499465522935955,
                  'NGC7496': 1.6527629311926655}

dict_psf_copt = {"NGC0628": 0.92,
                 "NGC1087": 0.92,
                 "NGC1365": 1.15,
                 "NGC1512": 1.25,
                 "NGC1672": 0.96,
                 "NGC2835": 1.15,
                 "NGC3627": 1.05,
                 "IC5332": 0.87,
                 "NGC4254": 0.89,
                 "NGC4535": 0.56,
                 "NGC5068": 1.04,
                 "NGC1300": 0.89,
                 "NGC1385": 0.67,
                 "NGC1433": 0.91,
                 "NGC1566": 0.80,
                 "NGC3351": 1.05,
                 "NGC4303": 0.78,
                 "NGC4321": 1.16,
                 "NGC7496": 0.89,
                }

dict_psf_IC5332 = {1: ['moffat', 0.595, 2.8, 6483.58, -3e-5],
                  2: ['moffat', 0.798, 2.8, 6483.58, -3e-5],
                  3: ['moffat', 0.667, 2.8, 6483.58, -3e-5],
                  4: ['moffat', 0.720, 2.8, 6483.58, -3e-5],
                  5: ['moffat', 0.751, 2.8, 6483.58, -3e-5]}

dict_psf_NGC0628 = {1: ['moffat', 0.733, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.771, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.733, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.841, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.742, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.621, 2.8, 6483.58, -3e-5],
                   7: ['moffat', 0.601, 2.8, 6483.58, -3e-5],
                   8: ['moffat', 0.695, 2.8, 6483.58, -3e-5],
                   9: ['moffat', 0.698, 2.8, 6483.58, -3e-5],
                   10: ['moffat', 0.752, 2.8, 6483.58, -3e-5],
                   11: ['moffat', 0.738, 2.8, 6483.58, -3e-5],
                   12: ['moffat', 0.664, 2.8, 6483.58, -3e-5]}

dict_psf_NGC1087 = {1: ['moffat', 0.695, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.788, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.829, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.630, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.845, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.626, 2.8, 6483.58, -3e-5]}

dict_psf_NGC1300 = {1: ['moffat', 0.781, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.660, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.752, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.813, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.538, 2.3, 6483.58, -3e-5],
                   6: ['moffat', 0.498, 2.3, 6483.58, -3e-5],
                   7: ['moffat', 0.631, 2.3, 6483.58, -3e-5],
                   8: ['moffat', 0.557, 2.3, 6483.58, -3e-5],
                   9: ['moffat', 0.577, 2.3, 6483.58, -3e-5],
                   10: ['moffat', 0.625, 2.3, 6483.58, -3e-5],
                   11: ['moffat', 0.536, 2.3, 6483.58, -3e-5],
                   12: ['moffat', 0.692, 2.3, 6483.58, -3e-5]}

dict_psf_NGC1365 = {1: ['moffat', 0.708, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.821, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.831, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.899, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.923, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.716, 2.8, 6483.58, -3e-5],
                   7: ['moffat', 0.640, 2.8, 6483.58, -3e-5],
                   8: ['moffat', 0.902, 2.8, 6483.58, -3e-5],
                   9: ['moffat', 0.901, 2.8, 6483.58, -3e-5],
                   10: ['moffat', 1.081, 2.8, 6483.58, -3e-5],
                   11: ['moffat', 0.583, 2.8, 6483.58, -3e-5],
                   12: ['moffat', 0.756, 2.8, 6483.58, -3e-5],
                   30: ['moffat', 0.823, 2.8, 6483.58, -3e-5]}

dict_psf_NGC1385 = {1: ['moffat', 0.489, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.379, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.568, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.694, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.591, 2.3, 6483.58, -3e-5]}

dict_psf_NGC1433 = {1: ['moffat', 0.701, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.630, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.831, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.601, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.628, 2.3, 6483.58, -3e-5],
                   6: ['moffat', 0.648, 2.3, 6483.58, -3e-5],
                   7: ['moffat', 0.702, 2.3, 6483.58, -3e-5],
                   8: ['moffat', 0.649, 2.3, 6483.58, -3e-5],
                   9: ['moffat', 0.709, 2.3, 6483.58, -3e-5],
                   10: ['moffat', 0.618, 2.3, 6483.58, -3e-5],
                   11: ['moffat', 0.609, 2.3, 6483.58, -3e-5],
                   12: ['moffat', 0.654, 2.3, 6483.58, -3e-5],
                   13: ['moffat', 0.646, 2.3, 6483.58, -3e-5],
                   14: ['moffat', 0.506, 2.3, 6483.58, -3e-5],
                   15: ['moffat', 0.642, 2.3, 6483.58, -3e-5]}

dict_psf_NGC1512 = {1: ['moffat', 0.732, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.847, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 1.176, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.800, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.683, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.699, 2.8, 6483.58, -3e-5],
                   7: ['moffat', 0.829, 2.8, 6483.58, -3e-5],
                   8: ['moffat', 0.927, 2.8, 6483.58, -3e-5],
                   30: ['moffat', 0.636, 2.8, 6483.58, -3e-5]}

dict_psf_NGC1566 = {1: ['moffat', 0.539, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.557, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.596, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.649, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.633, 2.3, 6483.58, -3e-5],
                   6: ['moffat', 0.723, 2.3, 6483.58, -3e-5],
                   7: ['moffat', 0.642, 2.3, 6483.58, -3e-5],
                   30: ['moffat', 0.641, 2.3, 6483.58, -3e-5]}

dict_psf_NGC1672 = {1: ['moffat', 0.651, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.887, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.729, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.645, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.798, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.773, 2.8, 6483.58, -3e-5],
                   7: ['moffat', 0.715, 2.8, 6483.58, -3e-5],
                   8: ['moffat', 0.683, 2.8, 6483.58, -3e-5]}

dict_psf_NGC2835 = {1: ['moffat', 0.775, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 1.080, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.932, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.673, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.849, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.709, 2.8, 6483.58, -3e-5],
                   30: ['moffat', 0.867, 2.8, 6483.58, -3e-5]}

dict_psf_NGC3351 = {1: ['moffat', 0.715, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.758, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.656, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.731, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.825, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.980, 2.8, 6483.58, -3e-5],
                   7: ['moffat', 0.839, 2.8, 6483.58, -3e-5],
                   8: ['moffat', 0.742, 2.8, 6483.58, -3e-5],
                   30: ['moffat', 0.612, 2.8, 6483.58, -3e-5]}

dict_psf_NGC3627 = {1: ['moffat', 0.675, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.781, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.980, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.754, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.798, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.767, 2.8, 6483.58, -3e-5],
                   7: ['moffat', 0.744, 2.8, 6483.58, -3e-5],
                   8: ['moffat', 0.814, 2.8, 6483.58, -3e-5]}

dict_psf_NGC4254 = {1: ['moffat', 0.633, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.573, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.588, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.609, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.815, 2.3, 6483.58, -3e-5],
                   6: ['moffat', 0.771, 2.3, 6483.58, -3e-5],
                   7: ['moffat', 0.619, 2.3, 6483.58, -3e-5],
                   8: ['moffat', 0.485, 2.3, 6483.58, -3e-5],
                   9: ['moffat', 0.454, 2.3, 6483.58, -3e-5],
                   10: ['moffat', 0.534, 2.3, 6483.58, -3e-5],
                   11: ['moffat', 0.444, 2.3, 6483.58, -3e-5],
                   12: ['moffat', 0.451, 2.3, 6483.58, -3e-5]}

dict_psf_NGC4303 = {1: ['moffat', 0.554, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.637, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.587, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.536, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.579, 2.3, 6483.58, -3e-5],
                   6: ['moffat', 0.533, 2.3, 6483.58, -3e-5],
                   7: ['moffat', 0.507, 2.3, 6483.58, -3e-5],
                   8: ['moffat', 0.612, 2.3, 6483.58, -3e-5],
                   9: ['moffat', 0.700, 2.3, 6483.58, -3e-5]}

dict_psf_NGC4321 = {1: ['moffat', 0.791, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.588, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.846, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.474, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.718, 2.3, 6483.58, -3e-5],
                   6: ['moffat', 0.463, 2.3, 6483.58, -3e-5],
                   7: ['moffat', 0.615, 2.3, 6483.58, -3e-5],
                   8: ['moffat', 0.997, 2.3, 6483.58, -3e-5],
                   9: ['moffat', 0.499, 2.3, 6483.58, -3e-5],
                   10: ['moffat', 0.673, 2.3, 6483.58, -3e-5],
                   11: ['moffat', 1.092, 2.3, 6483.58, -3e-5]}

dict_psf_NGC4535 = {1: ['moffat', 0.473, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.428, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.427, 2.3, 6483.58, -3e-5],
                   4: ['moffat', 0.469, 2.3, 6483.58, -3e-5],
                   5: ['moffat', 0.443, 2.3, 6483.58, -3e-5],
                   6: ['moffat', 0.433, 2.3, 6483.58, -3e-5]}

dict_psf_NGC5068 = {1: ['moffat', 0.669, 2.8, 6483.58, -3e-5],
                   2: ['moffat', 0.880, 2.8, 6483.58, -3e-5],
                   3: ['moffat', 0.962, 2.8, 6483.58, -3e-5],
                   4: ['moffat', 0.690, 2.8, 6483.58, -3e-5],
                   5: ['moffat', 0.622, 2.8, 6483.58, -3e-5],
                   6: ['moffat', 0.833, 2.8, 6483.58, -3e-5],
                   7: ['moffat', 0.773, 2.8, 6483.58, -3e-5],
                   8: ['moffat', 0.557, 2.8, 6483.58, -3e-5],
                   9: ['moffat', 0.523, 2.8, 6483.58, -3e-5],
                   10: ['moffat', 0.904, 2.8, 6483.58, -3e-5]}

dict_psf_NGC7496 = {1: ['moffat', 0.616, 2.3, 6483.58, -3e-5],
                   2: ['moffat', 0.813, 2.3, 6483.58, -3e-5],
                   3: ['moffat', 0.786, 2.3, 6483.58, -3e-5]}

dict_locals = locals()


src/pmuse_utils/phangs_muse_init.py
# -*- coding: utf-8 -*-

# -------------- General initialisation of muse quantities ---------
# This also serves to provide some useful structure for maps and dat
# EE - 2020. 
# v1.0 - Dec 2020
# ------------------------------------------------------------------

# -- Importing the usual packages. 
# Numpy, sys, copy, os
import numpy as np
import sys
from os.path import join as joinpath, isfile
import copy
import importlib
from astropy.io import fits as pyfits

# Importing the config parameters
from .phangs_muse_config import *

class Phangs_Muse_Info(object):
    """Main phangs class to gather some useful functions and info

    DR (str): data release number (e.g., 2.1)
    BB (str): WFI release number (e.g., 1.0)
    base_folder (str): folder top of tree
    """
    def __init__(self, **kwargs_phangs):

        # ============= FOLDERS =======================================
        # DR2 data folder
        print(f"INFO ------ we will be using data from DR {DR} --------")
        self.dict_phangs_keywords = define_phangs_keywords()
        for keyword in self.dict_phangs_keywords:
            setattr(self, keyword, kwargs_phangs.pop(keyword, self.dict_phangs_keywords[keyword]))

        self.dict_obs = {'WFI': ['WFI_BB', self.wfi_folder], 
                         'DUPONT': ['DUPONT_R', self.dp_folder]}
        
        # import sample and dictionaries
        pm_dict =  importlib.import_module(f'.dict.{dict_config_files[DR][0]}', package="pmuse_utils")
        psf_dict =  importlib.import_module(f'.dict.{dict_config_files[DR][1]}', package="pmuse_utils")

        for key in psf_dict.dict_locals:
            if key[0] != "_" and key != "dict_locals":
                setattr(self, key, getattr(psf_dict, key))
        for key in pm_dict.dict_locals:
            if key[0] != "_" and key != "dict_locals":
                setattr(self, key, getattr(pm_dict, key))

        # Phangs table
        self._version_phangs_table = "dummy"

    @property
    def nphangs(self):
        """Number of phangs targets
        """
        return len(self.phangs_sample)

    @property
    def nmuse(self):
        """Number of muse targets
        """
        return len(self.phangs_muse_sample)

    def _check_muse_targetname(self, targetname):
        """Check if target is in the muse sample
        """
        if targetname not in self.phangs_muse_sample:
            print(f"ERROR: {targetname} not in PHANGS MUSE sample [DR = {self.DR}]")
            return False
        return True

    def get_bbname(self, targetname, cut=False, size=5):
        """Get the reference image for a given galaxy
    
        Input
        -----
        targetname (str): name of the galaxy
        """
        if not self._check_muse_targetname(targetname): return ""

        if cut:
            add = f"_{size}x{size}"
        else:
            add = ""
        obs = self.dict_filter_for_alignment[targetname]
        return f"{self.dict_obs[obs][1]}{targetname}_Rc_flux_nosky{add}.fits"

    @property
    def phangs_table(self):
        version = self.version_phangs_table
        if self._version_phangs_table != version:
            phangs_tablename = joinpath(self.phangs_data_folder, 
                                        f"phangs_sample_table_{version}.fits")
            if isfile(phangs_tablename):
                self.phangs_tablename = phangs_tablename
                from astropy.table import Table
                self.version_phangs_table = copy.copy(self._version_phangs_table)
                self._phangs_table = Table.read(phangs_tablename)
            else:
                print(f"WARNING: file {phangs_tablename} does not exist - setting None")
                self._phangs_table = None
        return self._phangs_table

    @property
    def phangs_sample(self):
        return list(self.phangs_table['name'].data.astype(str))

    @property
    def muse_table(self):
        return self.phangs_table[self.phangs_table['survey_muse_status'] == 'released']

    def get_muse_pc_arcsec(self, targetname):
        dist = self.get_muse_param_1gal(targetname, 'dist')
        return dist * np.pi / 0.648

    def get_muse_param_1gal(self, targetname, param_name='dist'):
        """Get one param from the muse table and the value
        """
        if not self._check_muse_targetname(targetname): return None
        ind = np.argwhere(self.muse_table['name'].data.astype('str') == targetname.lower())[0][0]
        return self.muse_table[param_name][ind]

    def get_muse_param(self, param_name='dist'):
        """Get one param from the muse table and return dictionary
        """
        dict_param = {}
        for i in range(self.nmuse):
            dict_param[self.muse_table['name'][i].upper()] = self.muse_table[param_name][i]
        return dict_param

    def get_muse_fov(self, spaxel=0.2, unit="asec"):
        """Get the field of view from the dictionary of number
        of spaxels. Transform in kpc and asec for further usage.
        """
        if not hasattr(self, 'dict_muse_npix'):
            self._get_all_muse_npix()
        dist = self.get_muse_param('dist')
        dict_fov = {}
        ngal = len(dist)
        grid_asec = np.zeros((2, ngal))
        grid_kpc = np.zeros_like(grid_asec)
        for i, key in enumerate(dist):
            d = np.float(dist[key])
            fackpc = np.pi * d / 0.648 / 1000.
            nf = np.array(self.dict_muse_npix[key])
            asec = nf * spaxel
            kpc = asec * fackpc
            grid_asec[:,i] = [asec[0], asec[1]]
            grid_kpc[:,i] = [kpc[0], kpc[1]]
            dict_fov[key] = [asec, kpc]
    
        if unit == "asec" : return grid_asec
        elif unit == "kpc" : return grid_kpc
        else: return grid_asec

    def get_summary_table(surveys=['HST', 'HI', 'ASTROSAT', 'MUSE', 'ALMA']): 
        list_ok = ['released', 'observed_not_released']
    
        # select sample
        name = self.phangs_table['name']
        ra = self.phangs_table['orient_ra']
        dec = self.phangs_table['orient_dec']
        log_m = self.phangs_table['props_mstar']
        log_sfr = self.phangs_table['props_sfr']
    
        dict_surveys = {}
        for survey in surveys:
            test_ok = [self.phangs_table['survey_{survey}_status'] == ok for ok in list_ok]
            dict_surveys[survey] = np.argwhere(np.logical_or.reduce(test_ok))
    
        return name, ra, dec, log_m, log_sfr, dict_surveys

    def get_sdss(self, z_limit=0.05, folder=None, name='GSWLC-X2.dat.gz'):
        from astropy.io import ascii
        if folder is None: folder = self.phangs_data_folder
        tbl = ascii.read(joinpath(folder, name))
        log_m, log_sfr = tbl['col10'], tbl['col12']
        if z_limit is not None:
            z = tbl['col8']
            ind = np.where(z < z_limit)
        else:
            ind = np.where(log_sfr > -np.inf)
        return log_m[ind], log_sfr[ind], z_limit

    def get_muse_footprint(self, targetname):
        if not self._check_muse_targetname(targetname): return
        return muse_footprint(targetname, self.mask_folder)

    def get_map_suffix(self, targetname, version="native"):
        if version == "native":
            return "", "_MAPS"
        elif version == "copt":
            psf = self.dict_psf_copt[targetname]
            print(f"Copt version. For galaxy {targetname}, PSF is {psf}asec")
            return f"_copt_{psf}asec", f"-{psf}asec_MAPS"

    def get_muse_npix(self, targetname, ref_images_dict=default_white, 
                     version="native", verbose=False, **kwargs):
        """Just getting the shape of the reference images
        """
        suffix, mapsuffix = self.get_map_suffix(targetname, version)

        images_folder = self.dict_phangs_keywords[f"{version}_folders"][0]
        ref_imaname = (f"{images_folder}{targetname}_"
                     f"IMAGE_FOV_{ref_images_dict['white']}_WCS_Pall_mad{suffix}.fits")
        if not isfile(ref_imaname):
            if verbose:
                print(f"WARNING: ref image {ref_imaname} not found for {targetname}")
            return np.empty((0,0))

        return {targetname: np.array(pyfits.getdata(ref_imaname).shape)[::-1]}

    def _get_all_muse_npix(self, ref_images_dict=default_white, 
                           version="native", verbose=False):
        """Just getting the shapes of all reference images
        """
        self.dict_muse_npix = {}
        for targetname in self.phangs_muse_sample:
            self.dict_muse_npix.update(self.get_muse_npix(targetname, ref_images_dict, 
                                 version, verbose))


# Footprints
class Muse_Footprint(object):
    """Class to use the footprint of MUSE pointings
    It will be using the white_footprint.txt files for each pointing, which were
    derived using footprintfinder.py -d -t 1 {ima}
    """
    def __init__(self, targetname, mask_folder=""):
        self.target = targetname
        self.mask_folder = mask_folder

    def filename(self, pointing=1):
        if pointing not in self.pointings:
            print(f"File {name} not found in existing list. Please check folder")
            return None
        name = f"{self.mask_folder}{self.target}_mask_IMAGE_FOV_P{pointing:02d}_white_footprint.txt"
        return name

    @property
    def pointings(self):
        import re
        pointings = []
        for name in self._filenames:
            numpoint = np.int(re.findall(r'P\d\d', name)[0][1:])
            pointings.append(numpoint)
        return sorted(pointings)

    @property
    def _filenames(self, verbose=False):
        import glob
        nf = sorted(glob.glob(f"{self.mask_folder}{self.target}_mask_"
                            f"IMAGE_FOV_P??_white_footprint.txt"))
        if verbose:
            print(f"Found {len(nf)} pointings for {self.target}")
        return nf

    def get_footprint(self, pointing=1):
        # Opening the file
        poly = open(self.filename(pointing=pointing))
        # Filtering the first lines
        lines = poly.readlines()
        clines = [l for l in lines if l[0] != '#']

        # Creating the structures
        poly_pix = np.zeros((len(clines), 2))
        poly_wcs = np.zeros_like(poly_pix)
        k = 0
        for i in range(len(clines)):
            sl = clines[i].split()
            if len(sl) > 4:
                continue
            poly_pix[k, 0] = np.float(sl[0])
            poly_pix[k, 1] = np.float(sl[1])
            poly_wcs[k, 0] = np.float(sl[2])
            poly_wcs[k, 1] = np.float(sl[3])
            k += 1
        self.poly_pix = poly_pix[:k]
        self.poly_wcs = poly_wcs[:k]

        self.center_pix = [np.float(clines[1].split()[6]), np.float(clines[1].split()[7])]
        self.center_wcs = [np.float(clines[1].split()[8]), np.float(clines[1].split()[9])]

